### ISSUE TITLE
Memory leak with LayerNorm and 2nd derivatives (gradient penalty WGAN)

### ISSUE THREAD

**Issue by phil510** [pytorch/pytorch#30592]
## 🐛 Bug

There appears to be a memory leak when using LayerNorm with gradient penalty for WGAN. No error is ever returned, but the process is killed. Removing LayerNorm resolves the issue, so I believe that is the culprit, The issue is very similar to a previous issue with BatchNorm that was resolved: https://discuss.pytorch.org/t/memory-not-freeing-when-using-batchnormbackward/5998

## To Reproduce

Below is code to reproduce the issue. You may have to increase n depending on your machine's memory.

```
import torch
import torch.nn as nn

class DNN(nn.Module):
    def __init__(self):
        super().__init__()
        
        n = 512
        self.net = nn.Sequential(
            nn.Linear(3, n, bias = False),
            nn.LayerNorm(n), # removing this fixes the issue
            nn.LeakyReLU(0.2, inplace = True),
            nn.Linear(n, 1)
        )
            
    def forward(self, x):
        return self.net(x)

def calc_gradient_penalty(D, real_data, fake_data, device = None):
    if device is None:
        device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
        
    data_shape = real_data.shape
    
    eps = torch.rand(data_shape[0], *([1] * (len(data_shape) - 1)))
    eps = eps.expand(data_shape)
    eps = eps.to(device)
    
    interpolates = eps * real_data + (1 - eps) * fake_data
    interpolates.requires_grad = True
    assert (interpolates.requires_grad)

    D_interpolates = D(interpolates)

    gradients = torch.autograd.grad(outputs = D_interpolates, 
                                    inputs = interpolates,
                                    grad_outputs = torch.ones(D_interpolates.size()),
                                    create_graph = True,
                                    retain_graph = True, 
                                    only_inputs = True)[0]
    
    gradients = gradients.view(data_shape[0], -1)
    gradient_penalty = (gradients.norm(2, dim = 1) - 1.0).pow(2).mean()
    return gradient_penalty

D = DNN()

real_data = torch.rand(5, 3)
fake_data = torch.rand(5, 3)

gp = calc_gradient_penalty(D, real_data, fake_data)
gp.backward() # process terminates without error
```

## Environment

 - PyTorch Version (e.g., 1.0): 1.3.0
 - OS (e.g., Linux): Windows 10
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source): NA
 - Python version: 3.7
 - CUDA/cuDNN version: NA - CPU only for right now
 - GPU models and configuration:
 - Any other relevant information:

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @SsnL

**Comment by pietern** [comment]
Thanks for reporting. The BatchNorm leak you're referring to was fixed in #2326.

I briefly looked at the layer norm implementation but didn't see anything stand out...

cc @gchanan

**Comment by albanD** [comment]
Closing due to inactivity.
Feel free to re-open if you still see this with latest pytorch.
